### PR TITLE
feat: handle invalid areas / lanelets

### PR DESCRIPTION
**Pull request by ahmeddesokyebrahim**
## Description

This PR is one of group of PRs for providing a solution for the idea discussed [here](https://github.com/orgs/autowarefoundation/discussions/3275).

### Discussion
- [Discussion Link](https://github.com/orgs/autowarefoundation/discussions/3275)
### Issue 
- [Issue Link](https://github.com/autowarefoundation/autoware/issues/3308)
### Relevant PRs
  - [autoware_documentation](https://github.com/autowarefoundation/autoware-documentation/pull/337)
  - [autoware_common](https://github.com/autowarefoundation/autoware_common/pull/175)
  - [tier4/tier4_autoware_msgs - part I](https://github.com/tier4/tier4_autoware_msgs/pull/76) :heavy_check_mark: 
  - [tier4/tier4_autoware_msgs - part II](https://github.com/tier4/tier4_autoware_msgs/pull/87) :heavy_check_mark: 
  - [autoware_adapi_msgs](https://github.com/autowarefoundation/autoware_adapi_msgs/pull/27) :heavy_check_mark: 
  - [autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/235) :heavy_check_mark:
  - ~[tier4/tier4_ad_api_adaptor](https://github.com/tier4/tier4_ad_api_adaptor/pull/95)~
  
## Tests performed

- [x] Add a tag for a lenelet to mark it as invalid
- [x] Test the software using planning simulator with a vehicle approaching to this invalid lanelet to see that vehicle stops before entering the invalid lanelet area and ask the driver to take over
- [x] Test the software using planning simulator with a vehicle starts inside an invalid lanelet to see that vehicle does not drive in this area and ask the driver to take over

## Notes for reviewers

For compiling and testing this PR, all PRs in related links section need to be taken into consideration.
:exclamation: **Merge sequence can be found [here](https://github.com/orgs/autowarefoundation/discussions/3275#discussioncomment-5379038)** :exclamation: 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/rtc_manager_rviz_plugin/src/rtc_manager_panel.cpp
Comment (satoshi-ota) [inline common/rtc_manager_rviz_plugin/src/rtc_manager_panel.cpp:95]: This module will modify the output path shape? If it will modify only path velocity, this line diff is unnecessary.

File: planning/behavior_velocity_planner/invalid-lanelet-design.md
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/invalid-lanelet-design.md:26]: @ahmeddesokyebrahim Thanks for your contribution :clap: At first, could you fix these typos ?

Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/invalid-lanelet-design.md]: Thanks for your words Mr @satoshi-ota .
Resolved as well.

File: planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/manager.hpp
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/manager.hpp:1]: [nits]

```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/scene.hpp
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/scene.hpp:1]: [nits]

```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/util.hpp
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/util.hpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc., Leo Drive Teknoloji A.Ş.
```

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/debug.cpp
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/debug.cpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc., Leo Drive Teknoloji A.Ş.
```

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/manager.cpp
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/manager.cpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/util.cpp
Comment (satoshi-ota) [inline planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/util.cpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc., Leo Drive Teknoloji A.Ş.
```

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (mehmetdogru) [inline planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp:139]:
```suggestion
        const auto & stop_pose = op_stop_pose.get();
```

File: planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/scene.hpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/scene.hpp]: Fixed

File: planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/util.hpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/include/scene_module/invalid_lanelet/util.hpp]: Fixed

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/debug.cpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/debug.cpp]: Fixed

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp]: Fixed

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/util.cpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/util.cpp]: Fixed

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp]: Fixed

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp]: Fixed

File: planning/behavior_velocity_planner/invalid-lanelet-design.md
Comment (TakaHoribe) [inline planning/behavior_velocity_planner/invalid-lanelet-design.md:5]: The meaning of the `invalid lanelet` is ambiguous for newcomers. Would you add some examples or backgrounds of the `invalid lanelets`? (I know it is for the out of odd though.)

Or we can add a link if there is a description page for the `invalid lanelet`.

**Review by TakaHoribe — COMMENTED** [review]
@ahmeddesokyebrahim Thank you for the great work.

File: planning/behavior_velocity_planner/invalid-lanelet-design.md
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/invalid-lanelet-design.md:5]: Thanks for your comment @TakaHoribe san. 
Do you think adding the [discussion link](https://github.com/orgs/autowarefoundation/discussions/3275) will be sufficient for clearing ?

Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/invalid-lanelet-design.md:5]: @TakaHoribe :
I have added some background and examples in the readme file about invalid lanelets.
Please check and let me know if any modifications needed.

**Comment by TakaHoribe** [comment]
@ahmeddesokyebrahim Is this ready for a review? If so, could you fix the build failure on CI?

**Comment by ahmeddesokyebrahim** [comment]
Thanks for your comment @TakaHoribe san.

My PRs are all ready for review. The CI build/compilation failure I can't fix as my PR needs other PRs to have successful compilation. I have mentioned the merge sequence in this comment [here](https://github.com/orgs/autowarefoundation/discussions/3275#discussioncomment-5379038) and updated this PR description as well.

**Comment by TakaHoribe** [comment]
@ahmeddesokyebrahim I tested in my local. It seems to be working well.

image




However, when I move the vehicle forward and backward manually (by initialpose or external control command), the stop line moves following the vehicle. Would you fix this if it is not expected?

image

image

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (TakaHoribe) [inline planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp:39]: Would you think of simplification for this function? I guess it can be split into several functions.

**Review by TakaHoribe — COMMENTED** [review]
@ahmeddesokyebrahim The code overall looks good except for one comment about a long function. I also commented about the weird behavior found when I moved the vehicle manually after it stops at the invalid-lanelet stopline. Would you check them?

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp:38]: Thanks @TakaHoribe san for your review and providing the valuable comments.

For the `modifyPathVelocity` function simplification, I have worked on implementation refactoring and I hope it looks now better. Please have a look and let me know your thoughts.


For the the behavior of `stop line moves following the vehicle`, let me explain my understanding ...

- my understanding to the behavior of invalid lanelet module when it comes to `stopped state`, it should not go out of this state until the human driver takes over the driving responsibility and the autoware will be re-engaged (hence, re-initialized) and the virtual wall (stop line) will not be there as long as we are outside of an invalid lanelet after re-engaging . 

- As long as we are in an invalid lanelet, it is true that the stop pose is exactly in front of the vehicle so that it will prevent it from moving if the human driver re-engaged the autonomous mode.

- For the case of suddenly locating the vehicle in a new different start pose without turning off the system and turning on again, I guess this is not a real life situation. Do I need to handle this scenario or consider that autoware will be re-initialized in such case ?

Please correct my understanding about the above points if it is not true, and according to our discussion we can go forward if any changes needed.

**Comment by yukkysaito** [comment]
@ahmeddesokyebrahim Thank you for good PR.

I have an opinion on naming. :pray: 
The scene module should preferably be named in a way that is easy for the user to understand its function.
Lanelet is the name for lanelet2 and it is difficult to imagine its function from the user's perspective.
Example:
- no drivable lane/area

File: planning/behavior_velocity_planner/README.md
Comment (yukkysaito) [inline planning/behavior_velocity_planner/README.md:23]: @ahmeddesokyebrahim Thank you for good PR.

I have an opinion on naming. :pray: 
The scene module should preferably be named in a way that is easy for the user to understand its function.
Lanelet is the name for lanelet2 and it is difficult to imagine its function from the user's perspective.
Example:
- no drivable lane/area

Comment (ahmeddesokyebrahim) [reply planning/behavior_velocity_planner/README.md]: Thanks @yukkysaito san for your comment and contribution.
I have worked on changing the module name from `invalid_lanelet` to `no_drivable_lane` and reflected that to all [PRs](https://github.com/orgs/autowarefoundation/discussions/3275#discussioncomment-5379038).

As well, you can see a working [video demo](https://youtu.be/XKDQn2bcMPQ) after the name change and you can use [this map file](https://drive.google.com/file/d/1oH9Q0Sk9pm-ttF34Vx-bceQCx9Ja-xmS/view?usp=share_link) for testing if needed.

A reminder again : 
**:exclamation: Merge sequence can be found [here](https://github.com/orgs/autowarefoundation/discussions/3275#discussioncomment-5379038) :exclamation:**

Comment (yukkysaito) [reply planning/behavior_velocity_planner/README.md]: @ahmeddesokyebrahim 
Thanks for the correction. :+1: 
LGTM

**Review by yukkysaito — APPROVED** [review]
LGTM
@TakaHoribe I look like you have unresolved issue. Is it ok?

File: planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp
Comment (TakaHoribe) [reply planning/behavior_velocity_planner/src/scene_module/invalid_lanelet/scene.cpp]: > As long as we are in an invalid lanelet, it is true that the stop pose is exactly in front of the vehicle so that it will prevent it from moving if the human driver re-engaged the autonomous mode.

@ahmeddesokyebrahim Thank you, I have the same opinion.

What I tried is "after the vehicle is stopped by `invalid lanelet`, change to manual operation, move the vehicle backward, and then re-engage".
In this case, after the vehicle was moved backward up by manual driving and exited the `invalid lanelet`, it still continued to stop and could not be started by re-engaging. 
We can keep the stopline in front of the ego vehicle when it is in the `invalid lanelet`, but the stopline should not be out of the `invalid lanelet`.

Well, this is not a big issue and can be solved/discussed out of this PR. Let's merge the PR.

**Review by TakaHoribe — APPROVED** [review]

**Review by mehmetdogru — APPROVED** [review]
LGTM. @TakaHoribe could you take a look at this PR. Codeowner review is required.

**Review by mehmetdogru — APPROVED** [review]
LGTM

**Review by TakaHoribe — APPROVED** [review]
LGTM

**Review by TakaHoribe — APPROVED** [review]
LGTM

**Comment by mehmetdogru** [comment]
@TakaHoribe thank you for the approval however the PR still waits for a code owner review could anybody from your team approve as well? If not @xmfcx could you approve it as global-codeowner?

**Review by xmfcx — APPROVED** [review]

**Pull request merged**
